File: src/main/java/ca/warp7/tables/controller/main/TableController.java
package ca.warp7.tables.controller.main;

import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import javafx.fxml.FXML;
import javafx.geometry.Pos;
import javafx.scene.control.*;
import javafx.scene.control.cell.CheckBoxTableCell;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.control.cell.TextFieldTableCell;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.text.Font;
import javafx.util.Callback;

import java.util.Comparator;

public class TableController {

    @FXML
    TableView<Person> tableView;

    @FXML
    void initialize() {

        final Label label = new Label("Address Book");
        label.setFont(new Font("Arial", 20));

        final Label actionTaken = new Label();


        TableColumn<Person, String> firstNameCol = new TableColumn<>("First Name");
        firstNameCol.setMinWidth(100);
        firstNameCol.setCellValueFactory(new PropertyValueFactory<>("firstName"));
        firstNameCol.setCellFactory(param -> {
            TableCell<Person, String> cell = TextFieldTableCell.<Person>forTableColumn().call(param);
            cell.setEditable(true);
            return cell;
        });

        TableColumn<Person, String> lastNameCol = new TableColumn<>("Last Name");
        lastNameCol.setMinWidth(100);
        lastNameCol.setCellValueFactory(new PropertyValueFactory<>("lastName"));

        lastNameCol.setCellFactory(new Callback<TableColumn<Person, String>, TableCell<Person, String>>() {
            @Override
            public TableCell<Person, String> call(TableColumn<Person, String> param) {
                return new TableCell<Person, String>() {
                    @Override
                    protected void updateItem(String item, boolean empty) {
                        super.updateItem(item, empty);

                        setText(item);
                        if (!empty) {
                            String c = Integer.toHexString((int) (Math.random() * 255));
                            setStyle(String.format("-fx-background-color: #ff%s%s", c, c));
                        }
                    }
                };
            }
        });

        TableColumn<Person, String> emailCol = new TableColumn<>("Email");
        emailCol.setMinWidth(200);
        emailCol.setCellValueFactory(new PropertyValueFactory<>("email"));
        emailCol.setCellFactory(param -> new CheckBoxTableCell<Person, String>() {
            @Override
            public void updateItem(String item, boolean empty) {
                super.updateItem(item, empty);
                setText(item);
                setAlignment(Pos.CENTER_LEFT);
            }
        });

        TableColumn<Person, Person> btnCol = new TableColumn<>("Gifts");
        btnCol.setMinWidth(150);
        btnCol.setCellValueFactory(features -> new SimpleObjectProperty<>(features.getValue()));
        btnCol.setComparator(Comparator.comparing(Person::getLikes));

        tableView.setItems(FXCollections.observableArrayList(
                new Person("Jacob", "Smith", "jacob.smith@example.com", "The Button"),
                new Person("Isabella", "Johnson", "isabella.johnson@example.com", "The Button"),
                new Person("Ethan", "Williams", "ethan.williams@example.com", "The Button"),
                new Person("Emma", "Jones", "emma.jones@example.com", "The Button"),
                new Person("Michael", "Brown", "michael.brown@example.com", "The Button")

        ));
        tableView.setEditable(true);

        tableView.getSelectionModel().setCellSelectionEnabled(true);

        tableView.getColumns().add(firstNameCol);
        tableView.getColumns().add(lastNameCol);
        tableView.getColumns().add(emailCol);

        tableView.setOnKeyPressed(event -> tableView.getSortOrder().clear());

        btnCol.setCellFactory(new Callback<TableColumn<Person, Person>, TableCell<Person, Person>>() {

            @Override
            public TableCell<Person, Person> call(TableColumn<Person, Person> btnCol) {

                return new TableCell<Person, Person>() {

                    final ImageView buttonGraphic = new ImageView();
                    final Button button = new Button();

                    {
                        button.setGraphic(buttonGraphic);
                        button.setMinWidth(130);
                    }

                    @Override
                    public void updateItem(final Person person, boolean empty) {

                        super.updateItem(person, empty);
                        if (person != null) {
                            if ("fruit".equals(person.getLikes().toLowerCase())) {
                                button.setText("Buy fruit");
                                buttonGraphic.setImage(fruitImage);
                            } else {
                                button.setText("Buy coffee");
                                buttonGraphic.setImage(coffeeImage);
                            }

                            setGraphic(button);
                            button.setOnAction(event -> actionTaken.setText("Bought " + person.getLikes().toLowerCase() + " for: " + person.getFirstName() + " " + person.getLastName()));
                        } else {
                            setGraphic(null);
                        }
                    }
                };

            }
        });
    }

    @SuppressWarnings("ALL")
    public static class Person {

        private final SimpleStringProperty firstName;
        private final SimpleStringProperty lastName;
        private final SimpleStringProperty email;
        private final SimpleStringProperty likes;

        private Person(String fName, String lName, String email, String likes) {
            this.firstName = new SimpleStringProperty(fName);
            this.lastName = new SimpleStringProperty(lName);
            this.email = new SimpleStringProperty(email);
            this.likes = new SimpleStringProperty(likes);
        }

        public String getFirstName() {
            return firstName.get();
        }

        public void setFirstName(String fName) {
            firstName.set(fName);
        }

        public String getLastName() {
            return lastName.get();
        }

        public void setLastName(String fName) {
            lastName.set(fName);
        }

        public String getEmail() {
            return email.get();
        }

        public void setEmail(String fName) {
            email.set(fName);
        }

        public String getLikes() {
            return likes.get();
        }

        public void setLikes(String likes) {
            this.likes.set(likes);
        }
    }

    // icons for non-commercial use with attribution from: http://www.iconarchive.com/show/veggies-icons-by-iconicon/bananas-icon.html and http://www.iconarchive.com/show/collection-icons-by-archigraphs.html
    private final Image coffeeImage = new Image(
            "http://icons.iconarchive.com/icons/archigraphs/collection/48/Coffee-icon.png"
    );
    private final Image fruitImage = new Image(
            "http://icons.iconarchive.com/icons/iconicon/veggies/48/bananas-icon.png"
    );
}

File: UI/src/main/java/ca/warp7/rt/feature/views/TableController.kt
package ca.warp7.rt.feature.views

import javafx.beans.property.SimpleObjectProperty
import javafx.collections.FXCollections
import javafx.scene.control.*
import javafx.scene.control.cell.PropertyValueFactory
import javafx.scene.control.cell.TextFieldTableCell
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import javafx.scene.text.Font

class TableController {

    lateinit var tableContainer: VBox
    private val tableView = TableView<Person>()

    // icons for non-commercial use with attribution from: http://www.iconarchive.com/show/veggies-icons-by-iconicon/bananas-icon.html and http://www.iconarchive.com/show/collection-icons-by-archigraphs.html
    private val coffeeImage = Image(
            "http://icons.iconarchive.com/icons/archigraphs/collection/48/Coffee-icon.png"
    )
    private val fruitImage = Image(
            "http://icons.iconarchive.com/icons/iconicon/veggies/48/bananas-icon.png"
    )

    fun initialize() {

        VBox.setVgrow(tableView, Priority.ALWAYS)

        tableContainer.children.add(tableView)

        val label = Label("Address Book")
        label.font = Font("Arial", 20.0)

        val actionTaken = Label()


        val firstNameCol = TableColumn<Person, String>("First Name")
        firstNameCol.minWidth = 100.0
        firstNameCol.cellValueFactory = PropertyValueFactory("firstName")
        firstNameCol.setCellFactory { param ->
            val cell = TextFieldTableCell.forTableColumn<Person>().call(param)
            cell.isEditable = true
            cell
        }

        val lastNameCol = TableColumn<Person, String>("Last Name")
        lastNameCol.minWidth = 100.0
        lastNameCol.cellValueFactory = PropertyValueFactory("lastName")

        lastNameCol.setCellFactory {
            object : TableCell<Person, String>() {
                override fun updateItem(item: String?, empty: Boolean) {
                    super.updateItem(item, empty)
                    text = item
                    if (!empty) {
                        val c = Integer.toHexString((Math.random() * 255).toInt())
                        style = String.format("-fx-background-color: #ff%s%s", c, c)
                    }
                }
            }
        }

        val emailCol = TableColumn<Person, String>("Email")
        emailCol.minWidth = 200.0
        emailCol.cellValueFactory = PropertyValueFactory("email")

        val btnCol = TableColumn<Person, Person>("Gifts")
        btnCol.minWidth = 150.0
        btnCol.setCellValueFactory { features -> SimpleObjectProperty(features.value) }

        tableView.items = FXCollections.observableArrayList(
                Person("Jacob", "Smith", "jacob.smith@example.com", "The Button"),
                Person("Isabella", "Johnson", "isabella.johnson@example.com", "The Button"),
                Person("Ethan", "Williams", "ethan.williams@example.com", "The Button"),
                Person("Emma", "Jones", "emma.jones@example.com", "The Button"),
                Person("Michael", "Brown", "michael.brown@example.com", "The Button")

        )
        tableView.isEditable = true

        tableView.selectionModel.isCellSelectionEnabled = true

        tableView.columns.add(firstNameCol)
        tableView.columns.add(lastNameCol)
        tableView.columns.add(emailCol)

        tableView.setOnKeyPressed { tableView.sortOrder.clear() }

        btnCol.setCellFactory {
            object : TableCell<Person, Person>() {

                val buttonGraphic = ImageView()
                val button = Button()

                init {
                    button.graphic = buttonGraphic
                    button.minWidth = 130.0
                }

                public override fun updateItem(person: Person?, empty: Boolean) {
                    super.updateItem(person, empty)
                    if (person != null) {
                        if ("fruit" == person.getLikes().toLowerCase()) {
                            button.text = "Buy fruit"
                            buttonGraphic.image = fruitImage
                        } else {
                            button.text = "Buy coffee"
                            buttonGraphic.image = coffeeImage
                        }
                        graphic = button
                        button.setOnAction { actionTaken.text = "Bought " + person.getLikes().toLowerCase() + " for: " + person.getFirstName() + " " + person.getLastName() }
                    } else {
                        graphic = null
                    }
                }
            }
        }
    }
}
